File: packages/evals/tasks/bench/act/spif_in_osr.ts
import { z } from "zod";
import { defineBenchTask } from "../../../framework/defineTask.js";

export default defineBenchTask(
  { name: "spif_in_osr" },
  async ({ debugUrl, sessionUrl, stagehand, page, logger }) => {
    // this eval is designed to test whether stagehand can successfully
    // click inside a SPIF (same process iframe) that is inside an
    // OSR (open mode shadow) root

    try {
      await page.goto(
        "https://browserbase.github.io/stagehand-eval-sites/sites/spif-in-open-shadow-dom/",
      );
      await stagehand.act("click the button");

      const { data: extraction } = await stagehand.extract(
        "extract the entire page text",
        z.object({ extraction: z.string() }),
      );

      const pageText = extraction.extraction;

      if (pageText.includes("button successfully clicked")) {
        return {
          _success: true,
          message: `successfully clicked the button`,
          debugUrl,
          sessionUrl,
          logs: logger.getLogs(),
        };
      }
      return {
        _success: false,
        message: `unable to click on the button`,
        debugUrl,
        sessionUrl,
        logs: logger.getLogs(),
      };
    } catch (error) {
      return {
        _success: false,
        error: String(error),
        debugUrl,
        sessionUrl,
        logs: logger.getLogs(),
      };
    }
  },
);

File: packages/sdk-ruby/evals/tasks/act/spif_in_osr.rb
# frozen_string_literal: true

# Port of packages/evals/tasks/bench/act/spif_in_osr.ts

# this eval is designed to test whether stagehand can successfully
# click inside a SPIF (same process iframe) that is inside an
# OSR (open mode shadow) root
Evals.define_task("spif_in_osr") do |t|
  t.page.goto("https://browserbase.github.io/stagehand-eval-sites/sites/spif-in-open-shadow-dom/")
  t.stagehand.act("click the button")

  # v3 used schemaless extract; v4 requires a schema.
  # Single-word key to stay clear of the snake_case wire-casing bug (#14).
  extraction = t.stagehand.extract(
    "extract the entire page text",
    schema: {
      "type" => "object",
      "properties" => { "extraction" => { "type" => "string" } },
      "required" => ["extraction"],
      "additionalProperties" => false,
    },
  ).data

  page_text = extraction["extraction"]

  if page_text.include?("button successfully clicked")
    { _success: true, message: "successfully clicked the button" }
  else
    { _success: false, message: "unable to click on the button" }
  end
end
